File: 백준/Silver/16924. 십자가 찾기/십자가 찾기.py
dy = [-1,1,0,0]
dx = [0,0,-1,1]

def possible(cy,cx,sz):
    if target[cy][cx] == '.':
        return False

    for size in range(1,sz+1):
        for dir in range(4):
            ny = cy + dy[dir] * size
            nx = cx + dx[dir] * size

            if target[ny][nx] == '.':
                return False
    return True

def cover(sy,sx, sz):
    board[cy][cx] = '*'
    for size in range(1, sz+1):
        for dir in range(4):
            ny = cy + dy[dir] * size
            nx = cx + dx[dir] * size

            board[ny][nx] = '*'

def isEqual(target, board):
    for y in range(1,N+1):
        for x in range(1,M+1):
            if target[y][x] != board[y][x]:
                return False

    return True

if __name__ == '__main__':
    N,M = map(int, input().split())

    target = [ ['.'] * (M+1) for _ in range(N+1)]

    for y in range(1,N+1):
        line = input()
        for x in range(1, M+1):
            target[y][x] = line[x-1]

    maxSize = (min(N,M) - 1) // 2

    board = [['.'] * (M + 1) for _ in range(N + 1)]

    ans = []
    for sz in range(1, maxSize+1):
        for cy in range(1+sz, N - sz + 1):
            for cx in range(1 + sz, M - sz + 1):
                if possible(cy,cx, sz):
                    cover(cy,cx,sz)
                    ans.append((cy,cx,sz))

                    if isEqual(target, board):
                        print(len(ans))
                        for t in ans:
                            print(t[0], t[1], t[2])

                        exit()

    print(-1)

File: 백준/Silver/16924. 십자가 찾기/십자가 찾기.java
import java.util.*;
import java.io.*;

public class Main {
 
    public static void main(String[] args) throws IOException{
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	st = new StringTokenizer(br.readLine());
    	N = Integer.parseInt(st.nextToken());
    	M = Integer.parseInt(st.nextToken());
    	target = new char[N+1][M+1];
    	board = new char[N+1][M+1];
    	for(int y=1; y<=N; y++) {
    		String line = br.readLine();
    		for(int x=1; x<=M; x++) {
    			target[y][x] = line.charAt(x-1);
    			board[y][x] = '.';
    		}
    	}
    	
    	int maxSize = (Math.min(N, M) - 1) / 2;
    	
    	List<int[]> ans = new ArrayList<>();
    	
    	for(int size = 1; size<=maxSize;size++) {
    		for(int cy = 1 + size; cy<=N-size;cy++) {
    			for(int cx = 1 + size; cx<=M-size; cx++) {
    				if(possible(cy,cx,size)) {
    					cover(cy,cx,size);
    					ans.add(new int[] {cy,cx,size});
    					
    					if (isEqual(target, board)) {
    						StringBuilder sb = new StringBuilder();
    						sb.append(ans.size()).append('\n');
    						for(int[] t : ans) {
    							sb.append(t[0]+" "+t[1]+" "+t[2]).append('\n');
    						}
    						
    						System.out.println(sb);
    						System.exit(0);
    					}
    					
    				}
    			}
    		}
    	}
    	
    	System.out.println(-1);
    			
    			
    }
    
    static boolean possible(int cy, int cx, int sz) {
    	if (target[cy][cx] == '.')
    		return false;
    	
    	for(int size =1; size<=sz; size++) {
    		for(int dir = 0; dir< 4;dir++) {
    			int ny = cy + dy[dir] * size;
    			int nx = cx + dx[dir] * size;
    			
    			if(target[ny][nx] == '.')
    				return false;
    		}
    	}
    	
    	return true;
    }
    
    static void cover(int cy, int cx, int sz) {
    	board[cy][cx] = '*';
    	for(int size =1; size<=sz; size++) {
    		for(int dir = 0; dir< 4;dir++) {
    			int ny = cy + dy[dir] * size;
    			int nx = cx + dx[dir] * size;
    			
    			board[ny][nx] = '*';
    		}
    	}
    }
    
    static boolean isEqual(char target[][], char board[][]) {
    	for(int y=1; y<=N; y++) {
    		for(int x=1; x<=M; x++) {
    			if(target[y][x] != board[y][x])
    				return false;
    		}
    	}
    	
    	return true;
    }
    
    static int dy[] = {-1,1,0,0};
    static int dx[] = {0,0,-1,1};
    
    static StringTokenizer st;
    static int N,M;
    static char target[][], board[][];
}